Mathlib/RingTheory/Unramified/Field.lean

```diff
@@ -0,0 +1,212 @@
+    Function.Bijective (algebraMap K A) := by
+  have := finite_of_free (R := K) (S := A)
+  have : IsArtinianRing A := isArtinian_of_tower K inferInstance
+  have hA : IsNilpotent (LocalRing.maximalIdeal A) := by
+    rw [← LocalRing.jacobson_eq_maximalIdeal ⊥]
+    · exact IsArtinianRing.isNilpotent_jacobson_bot
+    · exact bot_ne_top
+  have : Function.Bijective (Algebra.ofId K (A ⧸ LocalRing.maximalIdeal A)) :=
+    ⟨RingHom.injective _, IsAlgClosed.algebraMap_surjective_of_isIntegral⟩
+  let e : K ≃ₐ[K] A ⧸ LocalRing.maximalIdeal A := {
+    __ := Algebra.ofId K (A ⧸ LocalRing.maximalIdeal A)
+    __ := Equiv.ofBijective _ this }
+  let e' : A ⊗[K] (A ⧸ LocalRing.maximalIdeal A) ≃ₐ[A] A :=
+    (Algebra.TensorProduct.congr AlgEquiv.refl e.symm).trans (Algebra.TensorProduct.rid K A A)
+  let f : A ⧸ LocalRing.maximalIdeal A →ₗ[A] A := e'.toLinearMap.comp (sec K A _)
+  have hf : (Algebra.ofId _ _).toLinearMap ∘ₗ f = LinearMap.id := by
+    dsimp [f]
+    rw [← LinearMap.comp_assoc, ← comp_sec K A]
+    congr 1
+    apply LinearMap.restrictScalars_injective K
+    apply _root_.TensorProduct.ext'
+    intros r s
+    obtain ⟨s, rfl⟩ := e.surjective s
+    suffices s • (Ideal.Quotient.mk (LocalRing.maximalIdeal A)) r = r • e s by
+      simpa [ofId, e']
+    simp [Algebra.smul_def, e, ofId, mul_comm]
+  have hf₁ : f 1 • (1 : A ⧸ LocalRing.maximalIdeal A) = 1 := by
+    rw [← algebraMap_eq_smul_one]
+    exact LinearMap.congr_fun hf 1
+  have hf₂ : 1 - f 1 ∈ LocalRing.maximalIdeal A := by
+    rw [← Ideal.Quotient.eq_zero_iff_mem, map_sub, map_one, ← Ideal.Quotient.algebraMap_eq,
+     algebraMap_eq_smul_one, hf₁, sub_self]
+  have hf₃ : IsIdempotentElem (1 - f 1) := by
+    apply IsIdempotentElem.one_sub
+    rw [IsIdempotentElem, ← smul_eq_mul, ← map_smul, hf₁]
+  have hf₄ : f 1 = 1 := by
+    obtain ⟨n, hn⟩ := hA
+    have : (1 - f 1) ^ n = 0 := by
+      rw [← Ideal.mem_bot, ← Ideal.zero_eq_bot, ← hn]
+      exact Ideal.pow_mem_pow hf₂ n
+    rw [eq_comm, ← sub_eq_zero, ← hf₃.pow_succ_eq n, pow_succ, this, zero_mul]
+  refine Equiv.bijective ⟨algebraMap K A, ⇑e.symm ∘ ⇑(algebraMap A _), fun x ↦ by simp, fun x ↦ ?_⟩
+  have : ⇑(algebraMap K A) = ⇑f ∘ ⇑e := by
+    ext k
+    conv_rhs => rw [← mul_one k, ← smul_eq_mul, Function.comp_apply, map_smul,
+      LinearMap.map_smul_of_tower, map_one, hf₄, ← algebraMap_eq_smul_one]
+  rw [this]
+  simp only [Function.comp_apply, AlgEquiv.apply_symm_apply, algebraMap_eq_smul_one,
+    map_smul, hf₄, smul_eq_mul, mul_one]
+
+theorem isField_of_isAlgClosed_of_localRing
+    [IsAlgClosed K] [LocalRing A] : IsField A := by
+  rw [LocalRing.isField_iff_maximalIdeal_eq, eq_bot_iff]
+  intro x hx
+  obtain ⟨x, rfl⟩ := (bijective_of_isAlgClosed_of_localRing K A).surjective x
+  show _ = 0
+  rw [← (algebraMap K A).map_zero]
+  by_contra hx'
+  exact hx ((isUnit_iff_ne_zero.mpr
+    (fun e ↦ hx' ((algebraMap K A).congr_arg e))).map (algebraMap K A))
+
+theorem isReduced_of_field :
```

Comment:
The name is strange, isn't it? There is no `field` assumption.

Reply:
The statement is "an unramified algebra over a field is reduced", so I gave it the name `FormallyUnramified.isReduced_of_field`. Do you have a better suggestion?

Reply:
Well, not really. Let's merge this, we can always change the name later.